### PR TITLE
Add null-safety support

### PR DESCRIPTION
**Pull request by Chever-John**
Branch master: 
    1. move to a runnable version.
Branch null_safety:
    1. support NullSafety 1.1

### CODE REVIEW

File: lib/data/blocs/bloc.dart [hunk 1]
```diff
@@ -1,49 +1,48 @@
...
-      [DateTime time, String config]) async {
+      [DateTime? time, String? config]) async {
     return await db.addDumpedRecord(DumpedRecordsCompanion().copyWith(
         data: Value(data),
         time: Value(time ?? DateTime.now()),
         config: Value(config ?? DEFAULT_CONFIG)));
   }
 
-  void updateDumpedRecordConfig(int id, String config) {
+  void updateDumpedRecordConfig(int? id, String config) {
```
Comment (Harry-Chen) [inline lib/data/blocs/bloc.dart:28]: Why bother using `int?` if you are writing `id!` immediately?

File: lib/data/blocs/bloc.dart [hunk 2]
```diff
@@ -1,49 +1,48 @@
...
-        id,
+        id!,
         DumpedRecordsCompanion(
           config: Value(config),
         ));
   }
 
-  Future<void> delDumpedRecord(int id) async {
-    await db.deleteDumpedRecord(id);
+  Future<void> delDumpedRecord(int? id) async {
```
Comment (Harry-Chen) [inline lib/data/blocs/bloc.dart:36]: Same problem as above

File: lib/data/blocs/provider.dart [hunk 3]
```diff
@@ -3,15 +3,15 @@ import 'package:flutter/material.dart';
 import 'bloc.dart';
 
 class BlocProvider extends InheritedWidget {
-  final NFSeeAppBloc bloc;
+  final NFSeeAppBloc? bloc;
```
Comment (Harry-Chen) [inline lib/data/blocs/provider.dart:6]: Why nullable?

File: lib/data/card.dart [hunk 4]
```diff
@@ -15,30 +15,38 @@ enum CardCategory {
...
 class CardData {
-  int id;
-  final CardCategory category;
-  String name;
-  final CardType cardType;
-  final String cardNo;
+  int? id;
+  final CardCategory? category;
+  String? name;
+  final CardType? cardType;
```
Comment (Harry-Chen) [inline lib/data/card.dart:21]: Shouldn't be nullable

File: lib/data/card.dart [hunk 5]
```diff
@@ -15,30 +15,38 @@ enum CardCategory {
 }
 
 class CardData {
-  int id;
-  final CardCategory category;
-  String name;
-  final CardType cardType;
-  final String cardNo;
+  int? id;
+  final CardCategory? category;
```
Comment (Harry-Chen) [inline lib/data/card.dart:19]: Shouldn't be nullable

File: lib/generated/intl/messages_all.dart [hunk 6]
```diff
@@ -19,12 +19,13 @@ import 'messages_en.dart' as messages_en;
 import 'messages_zh.dart' as messages_zh;
 
 typedef Future<dynamic> LibraryLoader();
+
```
Comment (Harry-Chen) [inline lib/generated/intl/messages_all.dart:22]: Is this file generated?

File: lib/generated/l10n.dart [hunk 7]
```diff
@@ -1,6 +1,7 @@
 // GENERATED CODE - DO NOT MODIFY BY HAND
 import 'package:flutter/material.dart';
 import 'package:intl/intl.dart';
+
```
Comment (Harry-Chen) [inline lib/generated/l10n.dart:4]: Is this file generated?

File: lib/main.dart [hunk 8]
```diff
@@ -96,17 +94,17 @@ class PlatformAdaptingHomePage extends StatefulWidget {
...
-  Exception error;
+  Exception? error;
   final GlobalKey<ScaffoldState> _scaffoldKey = new GlobalKey<ScaffoldState>();
 
-  PageController topController;
+  PageController? topController;
   int currentTop = 1;
 
-  NFSeeAppBloc get bloc => BlocProvider.provideBloc(context);
+  NFSeeAppBloc? get bloc => BlocProvider.provideBloc(context);
```
Comment (Harry-Chen) [inline lib/main.dart:107]: Shouldn't be nullable

File: lib/ui/card_physics.dart [hunk 9]
```diff
@@ -3,11 +3,12 @@ import 'package:flutter/cupertino.dart';
 const double CARD_WIDTH = (240 - 20) * (86.0 / 54) + 20;
 
 class CardPhysics extends ScrollPhysics {
-  final int cardCount;
-  const CardPhysics({ this.cardCount, parent }) : super(parent: parent);
+  final int? cardCount;
```
Comment (Harry-Chen) [inline lib/ui/card_physics.dart:6]: Shouldn't be nullable

File: lib/ui/custom_expansion_panel.dart [hunk 10]
```diff
@@ -78,29 +70,28 @@ class ExpansionPanel {
...
 
   /// The widget builder that builds the expansion panels' header.
-  final ExpansionPanelHeaderBuilder headerBuilder;
+  final ExpansionPanelHeaderBuilder? headerBuilder;
 
   /// The body of the expansion panel that's displayed below the header.
   ///
   /// This widget is visible only when the panel is expanded.
-  final Widget body;
+  final Widget? body;
```
Comment (Harry-Chen) [inline lib/ui/custom_expansion_panel.dart:84]: Shouldn't be nullable

File: lib/ui/custom_expansion_panel.dart [hunk 11]
```diff
@@ -78,29 +70,28 @@ class ExpansionPanel {
...
 
   /// Whether the panel is expanded.
   ///
   /// Defaults to false.
-  final bool isExpanded;
+  final bool? isExpanded;
```
Comment (Harry-Chen) [inline lib/ui/custom_expansion_panel.dart:89]: Shouldn't be nullable

File: lib/ui/custom_expansion_panel.dart [hunk 12]
```diff
@@ -78,29 +70,28 @@ class ExpansionPanel {
...
 
   /// Whether tapping on the panel's header will expand/collapse it.
   ///
   /// Defaults to false.
-  final bool canTapOnHeader;
-
+  final bool? canTapOnHeader;
```
Comment (Harry-Chen) [inline lib/ui/custom_expansion_panel.dart:94]: Shouldn't be nullable

File: lib/ui/settings.dart [hunk 13]
```diff
@@ -15,24 +14,24 @@ class SettingsAct extends StatefulWidget {
...
 
   @override
   void initState() {
     super.initState();
   }
 
   void _onTapDelete(BuildContext context, BuildContext outerCtx) async {
-    bool delRecords = false;
-    bool delScripts = false;
+    bool? delRecords = false;
```
Comment (Harry-Chen) [inline lib/ui/settings.dart:25]: Shouldn't be nullable!

File: pubspec.yaml [hunk 14]
```diff
@@ -4,34 +4,33 @@ description: An NFC data dumper.
 version: 2.1.0+5
 
 environment:
-  sdk: ">=2.6.0 <3.0.0"
+  sdk: ">=2.12.0 <3.0.0"
   flutter: ">=1.12.0 <2.0.0"
```
Comment (Harry-Chen) [inline pubspec.yaml:8]: Maybe should change to `>=2.0.0 <3.0.0`

Comment (Harry-Chen) [inline pubspec.yaml:8]: Maybe s
hould change to `>=2.0.0 <3.0.0`

**Review by Harry-Chen — CHANGES_REQUESTED** [review]
Mostly good. But there are more than expected nullable variables (maybe due to auto migration?). You should consider the semantics more carefully.

File: lib/data/blocs/bloc.dart (hunk 1 above)
Comment (Chever-John) [reply lib/data/blocs/bloc.dart:28]: Actually, there is no relation between `int?` and `id!`. 
`int ?` indicates that parameter `id` may be empty in function `void updateDumpedRecordConfig( )`.
`id!` indicates that parameter `id` may be empty and needed to be checked in function `db.writeDumpedRecord( )`.
Finally I reset the function `db.writeDumpedRecord( ) `, and conversation is resolved.

File: lib/data/card.dart (hunk 4 above)
Comment (Chever-John) [reply lib/data/card.dart:21]: There will have a conflict with [code](https://github.com/nfcim/nfsee/blob/master/lib/ui/home.dart#:~:text=if%20(pendingName%20%3D%3D%20%22%22)%20%7B-,data.name%20%3D%20null%3B,-%7D%20else%20%7B) in the home.dart, if we insist that `name` is nullable.

Comment (Chever-John) [reply lib/data/card.dart:21]: `cardType` is related to this [function](https://github.com/nfcim/nfsee/blob/master/lib/data/card.dart#:~:text=final%20cardType%20%3D%20getEnumFromString%3CCardType%3E(CardType.values%2C%20data%5B%27card_type%27%5D)%3B) that may have a null value return.

File: lib/ui/card_physics.dart (hunk 9 above)
Comment (Chever-John) [reply lib/ui/card_physics.dart:6]: Should be ? If it is set to non-nullable type, the following function calls will cause problems.

File: lib/ui/settings.dart (hunk 13 above)
Comment (Chever-John) [reply lib/ui/settings.dart:25]: Agree and set `v!` after.

**Review by Chever-John — COMMENTED** [review]
Almost of suggestions are useful for me. However, there are some problems I can't agree with u.

File: pubspec.yaml (hunk 14 above)
Comment (Chever-John) [reply pubspec.yaml]: I have been set flutter to ">=2.0.0 <3.0.0".

**Comment by Chever-John** [comment]
> Mostly good. But there are more than expected nullable variables (maybe due to auto migration?). You should consider the semantics more carefully.

Actually, the strategy I adopt is to change the code first. If I encounter problems, I will look at the automatically migrated code for inspiration. I will consider the semantics more carefully again after.

**Review by Harry-Chen — CHANGES_REQUESTED** [review]
Now please replace the i18n approach to that described at <https://flutter.dev/docs/development/accessibility-and-localization/internationalization>.

That is, you need to :

* remove all files under `lib/generated/`
* add dependency `flutter_localizations`
* Replace all usage of `S` to `AppLocalizations` (it is indeed too long, maybe you can find a way to alias it back to `S` if feasible)

**Review by Harry-Chen — APPROVED** [review]
LGTM now.

**Pull request merged**